### ISSUE TITLE
rota de fuga revisao 3

### ISSUE THREAD

**Issue by guilhermegandra** [guilhermegandra/Doutorado#3]
# ROTAS DE FUGA – REV 03
# - Entrada obrigatória: VIAS, EDIFICACOES, MANCHA_INUNDACAO (ZIP com SHP)
# - Entrada opcional: MDT (ZIP com GeoTIFF)
# - Saída: rotas.shp, saidas_afetadas.shp, pontos_encontro.shp,
#          imoveis_pontos.xlsx e resultados_rotas.zip

# -------------------------------------------------------------------
# Instalação de dependências (necessário apenas na primeira execução)
# -------------------------------------------------------------------
!apt-get install -y libspatialindex-dev > /dev/null 2>&1
!pip install -q geopandas shapely fiona rtree pyproj tqdm networkx scipy scikit-learn openpyxl pandas rasterio

# -------------------------------------------------------------------
# Imports
# -------------------------------------------------------------------
import os
import zipfile
import math
import time
import pandas as pd
import numpy as np

import geopandas as gpd
import networkx as nx
from shapely.geometry import Point, LineString
from shapely.ops import snap, split, unary_union, nearest_points
from tqdm import tqdm
from scipy.spatial import cKDTree
from sklearn.cluster import DBSCAN
from google.colab import files

import rasterio
from rasterio.warp import calculate_default_transform, reproject, Resampling

# -------------------------------------------------------------------
# Funções de controle de tempo
# -------------------------------------------------------------------
def start_stage(nome):
    print(f"\nETAPA: {nome}")
    return time.perf_counter()

def end_stage(t0):
    dt = time.perf_counter() - t0
    print(f"Tempo: {dt:.1f} s ({dt/60:.2f} min)")

# -------------------------------------------------------------------
# Função para upload e extração de ZIP com SHP
# -------------------------------------------------------------------
def upload_and_extract(label):
    print(f"\nEnvie o ZIP contendo o shapefile de: {label}")
    uploaded = files.upload()
    zip_name = list(uploaded.keys())[0]

    extract_dir = f"input_{label}"
    os.makedirs(extract_dir, exist_ok=True)

    with zipfile.ZipFile(zip_name, "r") as z:
        z.extractall(extract_dir)

    shp = [f for f in os.listdir(extract_dir) if f.lower().endswith(".shp")]

    if not shp:
        raise ValueError(f"O ZIP enviado para {label} não contém arquivo .shp")

    return os.path.join(extract_dir, shp[0])

# -------------------------------------------------------------------
# Upload dos dados obrigatórios
# -------------------------------------------------------------------
t0 = start_stage("Upload dos shapefiles ZIP (VIAS, EDIFICACOES, MANCHA_INUNDACAO)")
vias_path   = upload_and_extract("VIAS")
saidas_path = upload_and_extract("EDIFICACOES")
mancha_path = upload_and_extract("MANCHA_INUNDACAO")
end_stage(t0)

# -------------------------------------------------------------------
# Leitura dos shapefiles
# -------------------------------------------------------------------
t0 = start_stage("Leitura dos dados")
vias   = gpd.read_file(vias_path)
saidas = gpd.read_file(saidas_path)
mancha = gpd.read_file(mancha_path)
end_stage(t0)

# -------------------------------------------------------------------
# Garantir e ajustar CRS
# -------------------------------------------------------------------
def garantir_crs(gdf, nome):
    if gdf.crs is None:
        print(f"O arquivo '{nome}' não possui CRS.")
        epsg = int(input(f"Informe o EPSG original de {nome}: "))
        gdf = gdf.set_crs(epsg)
    return gdf

t0 = start_stage("Reprojeção para EPSG:31982 (SIRGAS 2000 / UTM 22S)")
vias   = garantir_crs(vias,   "vias").to_crs(31982)
saidas = garantir_crs(saidas, "edificacoes").to_crs(31982)
mancha = garantir_crs(mancha, "mancha").to_crs(31982)
end_stage(t0)

# -------------------------------------------------------------------
# Velocidade de deslocamento
# -------------------------------------------------------------------
t0 = start_stage("Parâmetros de deslocamento")
vel_kmh = float(input("Velocidade média (km/h): "))
vel_ms  = vel_kmh * 1000 / 3600
end_stage(t0)

# -------------------------------------------------------------------
# Limpeza e normalização da rede viária
# -------------------------------------------------------------------
t0 = start_stage("Limpeza e normalização da rede viária")

# Remove geometrias vazias e inválidas
vias = vias[~vias.geometry.is_empty & ~vias.geometry.isna()].copy()
vias["geometry"] = vias.geometry.apply(lambda g: g if g.is_valid else g.buffer(0))
vias = vias[vias.geometry.type.isin(["LineString", "MultiLineString"])].copy()

# Explode multilinhas
linhas = []
for geom in tqdm(vias.geometry, desc="Explodindo geometrias", ncols=80):
    if geom.geom_type == "LineString":
        linhas.append(geom)
    else:
        linhas.extend(list(geom.geoms))

# Cria união de todos os vértices
todos_pontos = [Point(c[:2]) for l in linhas for c in l.coords]
uniao_pontos = gpd.GeoSeries(todos_pontos, crs=vias.crs).unary_union

# Snap de cada linha para união de pontos (melhora conectividade)
snapped = []
for l in tqdm(linhas, desc="Aplicando SNAP", ncols=80):
    snapped.append(snap(l, uniao_pontos, tolerance=0.5))

# Split em interseções
vias_split = []
for i in tqdm(range(len(snapped)), desc="Split", ncols=80):
    sline = snapped[i]
    for j in range(len(snapped)):
        if i != j and sline.intersects(snapped[j]):
            try:
                sline = split(sline, snapped[j])
            except:
                pass

    if sline.geom_type in ["MultiLineString", "GeometryCollection"]:
        vias_split.extend([seg for seg in sline.geoms if seg.length > 0])
    else:
        if sline.length > 0:
            vias_split.append(sline)

end_stage(t0)

# -------------------------------------------------------------------
# Pergunta sobre MDT e carregamento opcional
# -------------------------------------------------------------------
t0 = start_stage("Pergunta sobre MDT (Modelo Digital do Terreno)")

usa_mdt    = input("A área analisada possui MDT? (Sim/Nao): ").strip().lower()
mdt_raster = None

if usa_mdt == "sim":
    print("\nEnvie o ZIP contendo o arquivo GeoTIFF do MDT:")
    uploaded = files.upload()
    zip_name = list(uploaded.keys())[0]

    extract_dir = "input_MDT"
    os.makedirs(extract_dir, exist_ok=True)

    with zipfile.ZipFile(zip_name, "r") as z:
        z.extractall(extract_dir)

    tif_list = [f for f in os.listdir(extract_dir)
                if f.lower().endswith(".tif") or f.lower().endswith(".tiff")]

    if not tif_list:
        raise ValueError("O ZIP enviado não contém arquivo .tif/.tiff de MDT.")

    mdt_path = os.path.join(extract_dir, tif_list[0])
    mdt_raster = rasterio.open(mdt_path)

    # Reprojeção do MDT para EPSG:31982, se necessário
    if mdt_raster.crs is None or mdt_raster.crs.to_epsg() != 31982:
        print("Reprojetando MDT para EPSG:31982...")
        dst_path = os.path.join(extract_dir, "mdt_31982.tif")

        transform, width, height = calculate_default_transform(
            mdt_raster.crs, "EPSG:31982",
            mdt_raster.width, mdt_raster.height,
            *mdt_raster.bounds
        )

        kwargs = mdt_raster.meta.copy()
        kwargs.update({
            "crs": "EPSG:31982",
            "transform": transform,
            "width": width,
            "height": height
        })

        with rasterio.open(dst_path, "w", **kwargs) as dst:
            for i in range(1, mdt_raster.count + 1):
                reproject(
                    source=rasterio.band(mdt_raster, i),
                    destination=rasterio.band(dst, i),
                    src_transform=mdt_raster.transform,
                    src_crs=mdt_raster.crs,
                    dst_transform=transform,
                    dst_crs="EPSG:31982",
                    resampling=Resampling.bilinear,
                )

        mdt_raster.close()
        mdt_raster = rasterio.open(dst_path)

end_stage(t0)

# -------------------------------------------------------------------
# Construção do grafo (com ou sem MDT)
# -------------------------------------------------------------------
t0 = start_stage("Construção do grafo viário (considerando MDT se existir)")

G = nx.Graph()
node_altitudes = {}

def obter_altitude(pt):
    """Interpolação no MDT. Se não houver MDT, retorna 0."""
    if mdt_raster is None:
        return 0.0
    try:
        for val in mdt_raster.sample([(pt.x, pt.y)]):
            return float(val[0])
    except Exception:
        return 0.0

for line in tqdm(vias_split, desc="Criando nós/arestas", ncols=80):
    coords = [tuple(c[:2]) for c in line.coords]

    # Garante altitude de cada nó
    for c in coords:
        if c not in node_altitudes:
            node_altitudes[c] = obter_altitude(Point(c[0], c[1]))

    # Cria arestas com peso = tempo ajustado (Naismith + Langmuir ajustados)
    for i in range(len(coords) - 1):
        p1 = coords[i]
        p2 = coords[i + 1]

        dist_m = math.dist(p1, p2)
        if dist_m == 0:
            continue

        # Tempo base (sem relevo) com a velocidade escolhida
        tempo = dist_m / vel_ms / 60.0

        if usa_mdt == "sim":
            alt1 = node_altitudes[p1]
            alt2 = node_altitudes[p2]
            ganho = alt2 - alt1  # positivo = subida, negativo = descida

            if ganho > 0:
                slope = ganho / dist_m  # declive positivo

                # Subidas leves, moderadas e fortes com penalidades diferentes
                if slope < 0.05:
                    tempo += (ganho / 100.0) * 8.0   # subida leve
                elif slope < 0.12:
                    tempo += (ganho / 100.0) * 12.0  # subida moderada
                else:
                    tempo += (ganho / 100.0) * 20.0  # subida forte / crítica

            elif ganho < 0:
                slope = abs(ganho) / dist_m  # declive em módulo

                # Descidas suaves: pequeno bônus
                if slope < 0.05:
                    tempo -= abs(ganho) * 0.015
                # Descidas moderadas: leve penalidade (cuidado maior)
                elif slope < 0.12:
                    tempo += abs(ganho) * 0.01
                # Descidas fortes: penalidade forte (perigoso / lento)
                else:
                    tempo += abs(ganho) * 0.05

        tempo = max(tempo, 0.0001)
        G.add_edge(p1, p2, weight=tempo)

# KD-Tree para localizar nós próximos
node_coords = np.array(list(G.nodes))
tree = cKDTree(node_coords)

def nearest_node(pt):
    _, idx = tree.query([pt.x, pt.y])
    return tuple(node_coords[idx])

end_stage(t0)

# -------------------------------------------------------------------
# Seleção das edificações afetadas pela mancha
# -------------------------------------------------------------------
t0 = start_stage("Identificação de edificações afetadas")
mancha_union    = mancha.geometry.unary_union
saidas_afetadas = gpd.sjoin(saidas, mancha, how="inner", predicate="intersects").reset_index(drop=True)
end_stage(t0)

# -------------------------------------------------------------------
# Função: encontrar nó viário fora da mancha
# -------------------------------------------------------------------
def nearest_node_outside_flood(point, tree, node_coords, mancha_union):
    for k in [1, 3, 5, 10, 20, 50, 100, 200]:
        _, idxs = tree.query([point.x, point.y], k=k)
        if k == 1:
            idxs = [idxs]
        for idx in idxs:
            cand = Point(node_coords[idx])
            if not cand.within(mancha_union):
                return tuple(node_coords[idx])
    raise ValueError("Nenhum nó viário fora da mancha encontrado.")

# -------------------------------------------------------------------
# Ponto global de encontro
# -------------------------------------------------------------------
t0 = start_stage("Cálculo do ponto global de encontro")

centro = saidas_afetadas.geometry.unary_union.centroid
if centro.within(mancha_union):
    pteo = nearest_points(centro, mancha_union.boundary)[1]
else:
    pteo = centro

dest_global  = nearest_node_outside_flood(pteo, tree, node_coords, mancha_union)
ponto_global = Point(dest_global[0], dest_global[1])

end_stage(t0)

# -------------------------------------------------------------------
# Rotas iniciais até o ponto global
# -------------------------------------------------------------------
t0 = start_stage("Cálculo das rotas iniciais")

rotas_ini   = []
tempos_ini  = []
rota_ok_ini = []

for idx, row in tqdm(saidas_afetadas.iterrows(),
                     total=len(saidas_afetadas),
                     desc="Rotas", ncols=80):
    try:
        start = nearest_node(row.geometry.centroid)
        tempo = nx.shortest_path_length(G, start, dest_global, weight="weight")
        tempo = math.ceil(tempo)
        path  = nx.shortest_path(G, start, dest_global, weight="weight")

        rotas_ini.append(LineString(path))
        tempos_ini.append(tempo)
        rota_ok_ini.append(True)
    except Exception:
        rotas_ini.append(None)
        tempos_ini.append(None)
        rota_ok_ini.append(False)

saidas_afetadas["tempo_min"] = tempos_ini
saidas_afetadas["rota_ok"]   = rota_ok_ini

end_stage(t0)

# -------------------------------------------------------------------
# Tempo máximo permitido
# -------------------------------------------------------------------
t0 = start_stage("Aplicação do tempo máximo permitido")
tempo_max = float(input("Tempo máximo permitido (min): "))
indices_excedidos = [i for i, t in enumerate(tempos_ini) if t is None or t > tempo_max]
end_stage(t0)

# -------------------------------------------------------------------
# DBSCAN para agrupar excedentes (500 m)
# -------------------------------------------------------------------
t0 = start_stage("Agrupamento DBSCAN para reduzir pontos de encontro")

pontos_teo = []
for i in indices_excedidos:
    g = saidas_afetadas.loc[i, "geometry"]
    c = g.centroid
    if c.within(mancha_union):
        c = nearest_points(c, mancha_union.boundary)[1]
    pontos_teo.append(c)

if pontos_teo:
    coords = np.array([[p.x, p.y] for p in pontos_teo])
    db = DBSCAN(eps=500, min_samples=1).fit(coords)
    labels   = db.labels_
    clusters = sorted(set(labels))
else:
    labels   = []
    clusters = []

end_stage(t0)

# -------------------------------------------------------------------
# Construção final das rotas considerando clusters
# -------------------------------------------------------------------
t0 = start_stage("Cálculo das rotas finais dos clusters")

rotas_final    = rotas_ini.copy()
tempos_final   = tempos_ini.copy()
id_ponto_final = [
    0 if tempos_ini[i] is not None and tempos_ini[i] <= tempo_max else None
    for i in range(len(saidas_afetadas))
]

pontos_encontro = [ponto_global]

for lab in clusters:
    cluster_indices = [indices_excedidos[k] for k, lb in enumerate(labels) if lb == lab]
    pts_cluster     = [pontos_teo[k]       for k, lb in enumerate(labels) if lb == lab]

    centro_cluster = unary_union(pts_cluster).centroid
    dest_cluster   = nearest_node_outside_flood(centro_cluster, tree, node_coords, mancha_union)
    pcluster       = Point(dest_cluster[0], dest_cluster[1])

    pontos_encontro.append(pcluster)
    pid = len(pontos_encontro) - 1

    for i in cluster_indices:
        try:
            start = nearest_node(saidas_afetadas.loc[i, "geometry"].centroid)
            tempo = nx.shortest_path_length(G, start, dest_cluster, weight="weight")
            tempo = math.ceil(tempo)
            path  = nx.shortest_path(G, start, dest_cluster, weight="weight")

            rotas_final[i]    = LineString(path)
            tempos_final[i]   = tempo
            id_ponto_final[i] = pid
        except Exception:
            rotas_final[i]    = None
            tempos_final[i]   = None
            id_ponto_final[i] = None

saidas_afetadas["tempo_min"] = tempos_final
saidas_afetadas["id_ponto"]  = id_ponto_final

end_stage(t0)

# -------------------------------------------------------------------
# Exportação dos resultados
# -------------------------------------------------------------------
t0 = start_stage("Exportação dos resultados")

os.makedirs("saida_colab/saida_final", exist_ok=True)

# Rotas
gpd.GeoDataFrame(
    {"tempo_min": tempos_final, "id_ponto": id_ponto_final},
    geometry=rotas_final,
    crs=vias.crs
).to_file("saida_colab/saida_final/rotas.shp")

# Edificações afetadas
saidas_afetadas.to_file("saida_colab/saida_final/saidas_afetadas.shp")

# Pontos de encontro
gpd.GeoDataFrame(
    {"id": list(range(len(pontos_encontro)))},
    geometry=pontos_encontro,
    crs=vias.crs
).to_file("saida_colab/saida_final/pontos_encontro.shp")

# XLSX de imóveis x pontos
df_xlsx = pd.DataFrame({
    "id_imovel": saidas_afetadas.index,
    "id_ponto_encontro": id_ponto_final,
    "tempo_min": tempos_final,
    "rota_ok": [r is not None for r in rotas_final]
})
df_xlsx.to_excel("saida_colab/saida_final/imoveis_pontos.xlsx", index=False)

# ZIP final
zip_name = "resultados_rotas.zip"
with zipfile.ZipFile(zip_name, "w", zipfile.ZIP_DEFLATED) as z:
    for root, _, files_list in os.walk("saida_colab"):
        for file in files_list:
            full_path = os.path.join(root, file)
            arcname   = os.path.relpath(full_path, "saida_colab")
            z.write(full_path, arcname)

files.download(zip_name)

end_stage(t0)

print("\nProcessamento finalizado com sucesso (REV 04).")
